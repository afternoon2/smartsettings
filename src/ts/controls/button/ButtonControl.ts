import { Control } from '../control/Control';
import {
  ButtonControlProps, ControlListener,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/button.sass';

export class ButtonControl extends Control {
  public controlElement: HTMLButtonElement;

  protected static defaultType: string = 'button';

  protected static template = (state: InternalState): string => `
    <button
      class="${Styles.button}"
      type="${state.type || ButtonControl.defaultType}"
      title="${state.name}"
      ${state.disabled ? 'disabled="true"' : ''}
    >
      ${state.text || ''}
    </button>
  `;

  private controlListener?: ControlListener;
  private sectionListener?: ControlListener;
  private panelListener?: ControlListener;

  constructor(props: ButtonControlProps) {
    super(props, ButtonControl.template);
    this.controlElement = this.element.querySelector('button') as HTMLButtonElement;
    this.controlListener = props.listener;
    this.sectionListener = props.sectionListener;
    this.panelListener = props.panelListener;
    this.bindActionListeners();
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('click', () => {
      if (this.controlListener) {
        this.controlListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'control'
        });
      }
      if (this.sectionListener) {
        this.sectionListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'section',
        });
      }
      if (this.panelListener) {
        this.panelListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'panel'
        });
      }
    });
  }
}
