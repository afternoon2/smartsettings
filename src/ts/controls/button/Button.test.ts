import { ButtonControl } from './ButtonControl';

let btn: ButtonControl;

const listener = jest.fn();

describe('Button control', () => {
  beforeEach(() => {
    btn = new ButtonControl({
      id: 'id',
      options: {
        name: 'Button',
        listener,
      },
      parentElement: document.body,
    });
  });

  test('If it matches the snapshot', () => {
    expect(btn.element).toMatchSnapshot();
  });

  test('Disable', () => {
    btn.disable();
    expect(btn.element).toMatchSnapshot();
  });

  test('Enable', () => {
    btn.disable();
    btn.enable();
    expect(btn.controlElement.hasAttribute('disabled')).toBe(false);
    expect(btn.element).toMatchSnapshot();
  });

  test('Invoke onClick and listener callback', () => {
    btn.controlElement.click();
    expect(listener).toHaveBeenCalledTimes(4);
  });
});
