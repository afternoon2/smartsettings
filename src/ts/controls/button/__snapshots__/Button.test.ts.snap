// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button control Disable 1`] = `
<div
  class="control disabled"
>
  <main
    class="control__body"
    id="id"
  >
    <div
      class="control__content control__content--button"
    >
      
    
      <button
        class="button"
        disabled="true"
        title="Button"
        type="button"
      >
        
      text
    
      </button>
      
  
    </div>
  </main>
</div>
`;

exports[`Button control Enable 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <div
      class="control__content control__content--button"
    >
      
    
      <button
        class="button"
        title="Button"
        type="button"
      >
        
      text
    
      </button>
      
  
    </div>
  </main>
</div>
`;

exports[`Button control If it matches the snapshot 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <div
      class="control__content control__content--button"
    >
      
    
      <button
        class="button"
        title="Button"
        type="button"
      >
        
      text
    
      </button>
      
  
    </div>
  </main>
</div>
`;
